### PR TITLE
Cleanup error handling to be clearer and remove unnecessary short-write ...

### PR DESCRIPTION
**Pull request by msolo-dropbox**
...check.

Based on comments from Sugu/Anthony more cleanup was warranted.

### CODE REVIEW

File: go/ioutil2/ioutil.go [hunk 1]
```diff
@@ -19,18 +18,17 @@ func WriteFileAtomic(filename string, data []byte, perm os.FileMode) error {
...
-		} else {
-			err = os.Chmod(f.Name(), perm)
-		}
+	if closeErr := f.Close(); err == nil {
+		err = closeErr
+	}
+	if permErr := os.Chmod(f.Name(), perm); err == nil {
+		err = permErr
 	}
+	// Any err should result in full cleanup.
```
Comment (enisoc) [inline go/ioutil2/ioutil.go]: What if the Rename fails?

File: go/ioutil2/ioutil_test.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package ioutil2
+
+import (
+	"os"
+	"testing"
+)
+
+func TestWrite(t *testing.T) {
+	fname := "/tmp/atomic-file-test.txt"
```
Comment (enisoc) [inline go/ioutil2/ioutil_test.go]: I think it's better to generate a random name. What if I happen to have a file named /tmp/atomic-file-test.txt that I always keep owned by root with 0400?

File: go/ioutil2/ioutil_test.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+	err := WriteFileAtomic(fname, []byte("test string\n"), 0664)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Remove(fname); err != nil {
```
Comment (enisoc) [inline go/ioutil2/ioutil_test.go]: Before removing, maybe the test could read back the file and check that all the data is there.

**Comment by msolo-dropbox** [comment]
PTAL

**Comment by alainjobart** [comment]
@enisoc @sougou

**Comment by enisoc** [comment]
LGTM. It still feels weird to do the "if err == nil, keep going" thing, but given the strange requirements we have around calling Sync and Close, anything else we do is going to be ugly as well. At least the happy path is non-nested now, which looks nicer.

It's also weird, although probably harmless, that we call Chmod even if the Write and/or Close failed.

**Comment by msolo-dropbox** [comment]
Yes, the is a pattern I've seen in the Go system library before. You will
do unnecessary work in the rare failure case, but the state tracking gets
even more complicated if you want to avoid that.

**Pull request merged**